homeassistant/components/demo/fan.py

```diff
@@ -166,7 +169,12 @@ def set_preset_mode(self, preset_mode: str) -> None:
             self._percentage = None
             self.schedule_update_ha_state()
         else:
-            raise ValueError(f"Invalid preset mode: {preset_mode}")
+            raise ServiceValidationError(
+                f"Invalid preset mode: {preset_mode}",
+                translation_key="invalid_fan_preset_mode",
+                translation_domain=DOMAIN,
+                translation_placeholders={"preset_mode": preset_mode},
+            )
```

Comment:
Hmm, I wonder if we should even add this here and instead update the entity component to handle this?

Reply:
`fan`component has `_valid_preset_mode_or_raise()` but it doesn't appear to be used in the service?!
I think some other components would also benefit from similar validation.

Reply:
Other chalenge is the UI. Possible we need an update for the Developers tools:

![afbeelding](https://github.com/home-assistant/core/assets/7188918/200238a5-9ae8-4fa3-8517-e51e3e5dbcc9)

Reply:
I have implemented `self._valid_preset_mode_or_raise(preset_mode)` for all occurrences.
We could also choose to always call this at entity level.
Could not find tests for it at `tests/components/fan/*`

Reply:
Right, why not

Reply:
Moved the preset_mode check to the entity platform now

Reply:
Updated fan entity component to raise on an invalid `preset_mode`.